art/estimators/speech_recognition/pytorch_deep_speech.py

```diff
@@ -0,0 +1,636 @@
+                    - Probability return is a tuple of (probs, sizes), where:
+                        - probs is the probability of characters of shape (nb_samples, seq_length, nb_classes).
+                        - sizes is the real sequence length of shape (nb_samples,).
+                    - Transcription return is a numpy array of characters. A possible example of a transcription return
+                      is `np.array(['SIXTY ONE', 'HELLO'])`.
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        x_ = x.copy()
+
+        # Put the model in the evaluation status
+        self._model.eval()
+
+        # Apply preprocessing
+        x_preprocessed, _ = self._apply_preprocessing(x_, y=None, fit=False)
+
+        # Transform x into the model input space
+        inputs, targets, input_rates, target_sizes, batch_idx = self._transform_model_input(x=x_preprocessed)
+
+        # Compute real input sizes
+        input_sizes = input_rates.mul_(inputs.size(-1)).int()
+
+        # Run prediction with batch processing
+        results = []
+        result_output_sizes = np.zeros(x_preprocessed.shape[0], dtype=np.int)
+        num_batch = int(np.ceil(len(x_preprocessed) / float(batch_size)))
+
+        for m in range(num_batch):
+            # Batch indexes
+            begin, end = (
+                m * batch_size,
+                min((m + 1) * batch_size, x_preprocessed.shape[0]),
+            )
+
+            # Call to DeepSpeech model for prediction
+            with torch.no_grad():
+                outputs, output_sizes = self._model(
+                    inputs[begin:end].to(self._device), input_sizes[begin:end].to(self._device)
+                )
+
+            results.append(outputs)
+            result_output_sizes[begin:end] = output_sizes.detach().cpu().numpy()
+
+        # Aggregate results
+        result_outputs = np.zeros(
+            (x_preprocessed.shape[0], result_output_sizes.max(), results[0].shape[-1]), dtype=np.float32
+        )
+
+        for m in range(num_batch):
+            # Batch indexes
+            begin, end = (
+                m * batch_size,
+                min((m + 1) * batch_size, x_preprocessed.shape[0]),
+            )
+
+            # Overwrite results
+            result_outputs[begin:end, : results[m].shape[1], : results[m].shape[-1]] = results[m].cpu().numpy()
+
+        # Rearrange to the original order
+        result_output_sizes_ = result_output_sizes.copy()
+        result_outputs_ = result_outputs.copy()
+        result_output_sizes[batch_idx] = result_output_sizes_
+        result_outputs[batch_idx] = result_outputs_
+
+        # Check if users want transcription outputs
+        transcription_output = kwargs.get("transcription_output")
+
+        if transcription_output is None or transcription_output is False:
+            return result_outputs, result_output_sizes
+
+        # Now users want transcription outputs
+        # Hence first create a decoder
```

Comment:
Can we create the decoder in way that allows is to reuse it in a second call to `predict`?Should we create the decoder in `__init__`?

Reply:
Yes, it is now put in __init__.